app/scripts/controllers/user-storage/encryption.ts

```diff
@@ -23,14 +36,57 @@ function bytesToUtf8(byteArray: Uint8Array) {
   return decoder.decode(byteArray);
 }
 
+/**
+ * In Memory Cache of the derived key from a given salt.
+ * We can tidy this to make it reusable with future versions and KDF functions
+ */
+const inMemCachedKDF: Map<string, Uint8Array> = new Map();
+
+type CachedEntry = {
+  salt: Uint8Array;
+  key: Uint8Array;
+};
+const getCachedKeyBySalt = (salt: Uint8Array): CachedEntry | undefined => {
+  const base64Salt = byteArrayToBase64(salt);
+  const cachedKey = inMemCachedKDF.get(base64Salt);
+  if (!cachedKey) {
+    return undefined;
+  }
+
+  return {
+    salt,
+    key: cachedKey,
+  };
+};
+const getAnyCachedKey = (): CachedEntry | undefined => {
+  const cachedEntry = Array.from(inMemCachedKDF.entries()).at(0);
+  if (!cachedEntry) {
+    return undefined;
+  }
+
+  const bytesSalt = base64ToByteArray(cachedEntry[0]);
+  return {
+    salt: bytesSalt,
+    key: cachedEntry[1],
+  };
+};
+const setCachedKey = (salt: Uint8Array, key: Uint8Array): void => {
```

Comment:
In case anyone else is wondering how more than one key would end up here - note that memory cache can be lost.
If you encrypt something, send it, restart the browser, encrypt something again and then fetch previously encrypted value - you have 2 keys.

Could there be more? Yes. mobile app and portfolio could do he same on various features' user data, so the maximum number of active keys is 1+number of features that can be stored.

As a result, if you keep your app alive with this implementation for a long period of time while other instances keep resetting, it could accumulate a long list of keys. Restarting fixes it tho.

(could) To avoid this rare memory leak scenario we might consider adding an if to setCachedKey to check if inMemCachedKDF.size is above a reasonable maximum (20? 100?) and clear cache if so to prevent it growing indefinitely.
The likelihood of that ever happening feels low tho.

Reply:
After some testing & profiling, our JS heap peaks at 222MB (including other processes and services). And memory snapshots show an increase between 20-40mb which shoots back down (including other processes and services running in background)

Reply:
Yep very low.

Good idea and set a limit of 10 cached keys. We can change this in the future if we ever need more.